Disallow sending CUDA tensors over RPC for current RPC agents.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33604 Disallow sending CUDA tensors over RPC for current RPC agents.**

For our current RPC agents, this PR disallows sending CUDA tensors
over RPC and asks users to copy them explicitly to CPU. Currently, this seems
to be the easiest contract to guarantee for our current RPC agents, otherwise
if we do support this transparently it gets a little tricky in terms of whether
a CUDA tensor on the client should be sent to CPU/GPU of the remote end and
also which GPU device on the remote end.

In the future, the TensorPipe RPC agent can have its own specific handling of
CUDA tensors.

Closes https://github.com/pytorch/pytorch/issues/28881

Differential Revision: [D20020183](https://our.internmc.facebook.com/intern/diff/D20020183/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20020183/)!

pritamdamania committed:
Disallow sending CUDA tensors over RPC for current RPC agents.

For our current RPC agents, this PR disallows sending CUDA tensors
over RPC and asks users to copy them explicitly to CPU. Currently, this seems
to be the easiest contract to guarantee for our current RPC agents, otherwise
if we do support this transparently it gets a little tricky in terms of whether
a CUDA tensor on the client should be sent to CPU/GPU of the remote end and
also which GPU device on the remote end.

In the future, the TensorPipe RPC agent can have its own specific handling of
CUDA tensors.

Closes https://github.com/pytorch/pytorch/issues/28881

Differential Revision: [D20020183](https://our.internmc.facebook.com/intern/diff/D20020183/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20020183/)!

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -427,7 +427,27 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+              for (const auto& tensor : respMessage.tensors()) {
+                if (!tensor.device().is_cpu()) {
+                  errorMsg = c10::str(
+                      "ProcessGroup RPC backend only supports",
```
would it be worth adding tensor.device().str() to the errorMsg, so we know which non-cpu device the tensor is from?

jjlilley reviewed (commented):
Checking to make sure I understand:

 1) currently, GPU tensors will work transparently (?), though not necessarily performantly, since we need to copy to the cpu first
   (i.e. I see the .cpu() call in pickler.cpp:548, and the BINPERSID handler in unpickler.cpp)
    That said, I've never tried using the support, so I don't know if it actually works.
    If it doesn't work, my position below is different :)

 2) this change is intentionally breaking that support.

If that's the case - architecturally, are we sure we want to do this?

Realistically, prod-ready TensorPipes is still a ways into the future. And in the meantime, this disables a potentially convenient feature.
 
Maybe we could make this just a configuration option (i.e. .disallowNonCPUTransfers)?
And/or spam with LOG_FIRST_N(WARNING) if they're rpcing GNU tensors?

pritamdamania87 commented:
@jjlilley Yes this works right now, not because we designed it this way but just as a matter of chance. Although, supporting something like this as a first class citizen for the RPC layer is tricky (and thats why we have TensorPipe). For example, today if I have a Tensor on GPU8 on Node A, the pickler will try to create the Tensor on GPU8 on Node B when you send it over RPC. Although, there might be only two GPUs on the remote end or it might not even have a GPU.

Instead of dealing with these cases in our current RPC agents, I feel this is a much simpler contract we can provide to users and in the future allow such optimizations via TensorPipe.

jjlilley commented:
> Yes this works right now, not because we designed it this way but just as a matter of chance.

@pritamdamania87 

1) So, end-users can make things continue to work by just calling .cpu() on one side and .cuda() on the other? 
   * In that case, maybe it's not so bad, though it feels like it's encouraging the wrong thing from a layering perspective - e.g. code ends up littered with .cpu()/.cuda() calls that become unnecessary (but alas are never culled in practice) when the new "upgraded" default rpc agent ends up doing the right thing for free.
   * I'm also concerned about abstractions like rref, where it's maybe less clear how things would work.

2) I completely agree that something like TensorPipes should provide a far better future experience for transporting multi-device tensors. But TensorPipes is still vaporware/demoware now. 
  And even after being code complete, it will take a bit of time to flush out the bugs.

Until then, it may be a bit heavy-handed to completely disable a working and potentially useful OSSed feature without providing a working alternative. That said, minimally, it wouldn't be a bad idea to add a performance-oriented LOG_FIRST_N(WARNING), as well as documentation saying that it's not a great idea.

3) btw, your "just a matter of chance" comment might take things a bit far. :) Folks like myself who worked on rpc_agent were definitely aware that the pickler gave us this functionality for free, even if it wasn't necessarily a first class feature.

pritamdamania87 commented:
> So, end-users can make things continue to work by just calling .cpu() on one side and .cuda() on the other?

Yes

> Until then, it may be a bit heavy-handed to completely disable a working and potentially useful OSSed feature without providing a working alternative. That said, minimally, it wouldn't be a bad idea to add a performance-oriented LOG_FIRST_N(WARNING), as well as documentation saying that it's not a great idea.

In OSS, the distributed RPC framework is explicitly marked as an experimental feature so that we can make such breaking changes. In PyTorch 1.5 (expected around mid March), we plan to remove this experimental tag and as a result we need a consistent contract for how we handle CUDA tensors. I would prefer to start with something simple where we only allow CPU tensors. If we allow CUDA tensors from the get go we need to handle all the edge cases (ex: no gpus on the remote end, different number of gpus on the remote end) across both of our RPC agents and probably need to also support future requests where users might want additional optimizations in terms of how we handle CUDA tensors. At some point in the future if there is a use case where calling .cpu() on one side and .cuda() on the other side is a perf problem and there is some reason we can't use TensorPipe either, then we can evaluate if we want this functionality in other RPC agents.

> btw, your "just a matter of chance" comment might take things a bit far. :) Folks like myself who worked on rpc_agent were definitely aware that the pickler gave us this functionality for free, even if it wasn't necessarily a first class feature.

My bad, what I meant was that when we designed the RPC API we didn't think of CUDA tensors at all since the initial use case was CPU only and as a result we didn't document the behavior for CUDA tensors, neither did we have any unit tests verifying the intended behavior.

mrshenli commented:
This change looks OK to me. Either we disallow it, or we clearly state we would preserve device affinity and will propagate error like, no such device if the counterpart device is missing on the destination.

Just curious, did any one complained about the current behavior?

jjlilley commented:
I suspect that the motivation is simply to tighten up the RpcAgent api contract before PyTorch 1.5 branch is cut, so that people don't rely on something we didn't set out to support.

If consensus is to merge this, I don't have a big objection. But my concern is unintended consequences with respect to RRef handling. Particularly, Tensors can be anywhere in the cluster and usable to somebody with the RRef, due to rpc/pickler magic. But now if they happen to be on the GPU, the abstraction now seems like it could break with this change. If that's not the case, great - but if it is, let's avoid accidentally breaking things.

pritamdamania87 commented:
> or we clearly state we would preserve device affinity and will propagate error like, no such device if the counterpart device is missing on the destination.

My concern with this approach is that the error would likely happen inside the jit::pickler and its not something we can easily "detect" and handle at the RPC layer to report an appropriate exception to the user.

> Just curious, did any one complained about the current behavior?

No, but as Jeremy mentioned, we need a clear contract for PyTorch 1.5

> If consensus is to merge this, I don't have a big objection. But my concern is unintended consequences with respect to RRef handling. Particularly, Tensors can be anywhere in the cluster and usable to somebody with the RRef, due to rpc/pickler magic. But now if they happen to be on the GPU, the abstraction now seems like it could break with this change. If that's not the case, great - but if it is, let's avoid accidentally breaking things.

The case you mentioned regarding RRefs would break with this PR, although I'm not sure if anyone is relying on this right now.


My concern with the behavior we have today is that it only works if both nodes have same number of GPUs. For example, if I send a GPU Tensor to a node that doesn't have a GPU, the user gets a cryptic error message:

```
RuntimeError: worker0: Error in response from worker1: std::runtime_error: cuda runtime error (3) : This binary is linked with CUDA lazy stubs and underlying .so files were not loaded. CUDA functionality is disabled. Set env variable CUDA_LAZY_DEBUG to get messages during startup. at caffe2/aten/src/THC/THCGeneral.cpp:51
```

mrshenli commented:
That error message indeed looks terrible. 

> My concern with this approach is that the error would likely happen inside the jit::pickler and its not something we can easily "detect" and handle at the RPC layer to report an appropriate exception to the user.

Looks like we can detect it in our [`wireDeserialize`](https://github.com/pytorch/pytorch/blob/dbe850af5b941c6f0ae1ed50e7d6b779db55f5af/torch/csrc/distributed/rpc/utils.cpp#L294) function?

pritamdamania87 commented:
> Looks like we can detect it in our wireDeserialize function?

IIUC, the unpickler needs to run (https://github.com/pytorch/pytorch/blob/dbe850af5b941c6f0ae1ed50e7d6b779db55f5af/torch/csrc/distributed/rpc/utils.cpp#L336), before we get any Tensors (and any device information). But the unpickler itself would throw a runtime_error as I've mentioned above. One option would be to detect this in the pickler and then throw a custom exception that we can catch and process on the RPC side.

Although, I do feel the contract in this PR much simpler to support in general. If you feel strongly we should preserve the GPU to GPU conversion we have in current RPC agents, I'll try to add these checks to the pickler and handle it on the RPC side.

jjlilley commented:
> > Looks like we can detect it in our wireDeserialize function?
> 
> IIUC, the unpickler needs to run (
> 
> https://github.com/pytorch/pytorch/blob/dbe850af5b941c6f0ae1ed50e7d6b779db55f5af/torch/csrc/distributed/rpc/utils.cpp#L336
> 
> ), before we get any Tensors (and any device information). But the unpickler itself would throw a runtime_error as I've mentioned above. One option would be to detect this in the pickler and then throw a custom exception that we can catch and process on the RPC side.

Right, the Unpickler sees a Device from the encoded stream, and then effectively calls 
  tensor = tensor.to(device, tensor.scalar_type())

with the provided device to move to the appropriate location.

I imagine that if we suggested a fairly clean callback insertion point here, we could remap if it's needed.

Wondering aloud: would it be acceptable to specify as a contract that:
  - RpcAgent fully supports cpu tensors
  - RpcAgent with Gpu tensors is experimental and subject to some limitations. Particularly, the transfer only works if the number of devices matches on both nodes, and goes through the CPU, so isn't optimally efficient.

> Although, I do feel the contract in this PR much simpler to support in general. If you feel strongly we should preserve the GPU to GPU conversion we have in current RPC agents, I'll try to add these checks to the pickler and handle it on the RPC side.

pritamdamania87 commented:
@jjlilley @mrshenli Got a chance to get back to this PR and after some thought I still do have concerns supporting transparent GPU copies in our current RPC agents. For example, if we have two processes on the same node where process 1 is responsible for GPU1 and process 2 is responsible for GPU2. It would be really weird behavior when process 1 sends a tensor on GPU1 to process 2. The tensor would end up on GPU1 itself.

I do feel we should stick to a much simpler contract here and just allow CPU tensors instead of dealing with all these edge cases. Even in the RPC-TensorPipe integration, the plan is to default always to CPU memory if no placement information is provided.

jjlilley commented:
Hi Pritam:

I don't completely mind if we merge. - That said:
 - There might well be unforseen consequences (e.g. what happens to rrefs that reference gpu tensors?  It seems there could be some situations that were [sort of] working, that would be broken by this)

 - Just saying - we've had many regressions in the codebase in the past month. Disabling existing functionality (rather than just saying "it's still experimental with gpu-based tensors") because we aren't sure about our future support plans feels a bit low-pri in the scheme of things.

pritamdamania87 commented:
> There might well be unforseen consequences (e.g. what happens to rrefs that reference gpu tensors? It seems there could be some situations that were [sort of] working, that would be broken by this)

I agree rrefs pointing to GPU tensors would be broken and this might break other things for users. Although, those can be fixed by doing explicit copies on the user side. For rref, instead of doing rref.to_here() on the GPU tensor, you can call a remote method that looks at the local value of the rref and returns the data to you.

> Just saying - we've had many regressions in the codebase in the past month. Disabling existing functionality (rather than just saying "it's still experimental with gpu-based tensors") because we aren't sure about our future support plans feels a bit low-pri in the scheme of things.

As per my understanding, the only use case we have internally for GPU tensors is Serhat's work. I checked with Serhat and he mentioned we don't currently rely on this functionality and instead do explicit cpu to gpu copies on the user side.

jjlilley commented:
Thinking about this, and review velocity:
1 It's probably best to up/down this change, rather than drag it on further.
2 I'm not necessarily a big fan, but it's not unreasonable thing to do. ("disagree and commit", courtesy of Amazon)
3 So, I'm ok with it, as long as tests pass, and there's a little bit more info in the error message (e.g. maybe the device() name, and maybe the tensor.numel()) to make it less cryptic.

Thanks for looking into this issue!

pritamdamania committed:
Update on "Disallow sending CUDA tensors over RPC for current RPC agents."


For our current RPC agents, this PR disallows sending CUDA tensors
over RPC and asks users to copy them explicitly to CPU. Currently, this seems
to be the easiest contract to guarantee for our current RPC agents, otherwise
if we do support this transparently it gets a little tricky in terms of whether
a CUDA tensor on the client should be sent to CPU/GPU of the remote end and
also which GPU device on the remote end.

In the future, the TensorPipe RPC agent can have its own specific handling of
CUDA tensors.

Closes https://github.com/pytorch/pytorch/issues/28881

Differential Revision: [D20020183](https://our.internmc.facebook.com/intern/diff/D20020183/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20020183/)!

[ghstack-poisoned]

pritamdamania87 commented:
Added tensor.device() to the error message.

jjlilley commented:
LGTM!
Though looks like there's a merge conflict? (at least on Phab side?)

pritamdamania committed:
Update on "Disallow sending CUDA tensors over RPC for current RPC agents."


For our current RPC agents, this PR disallows sending CUDA tensors
over RPC and asks users to copy them explicitly to CPU. Currently, this seems
to be the easiest contract to guarantee for our current RPC agents, otherwise
if we do support this transparently it gets a little tricky in terms of whether
a CUDA tensor on the client should be sent to CPU/GPU of the remote end and
also which GPU device on the remote end.

In the future, the TensorPipe RPC agent can have its own specific handling of
CUDA tensors.

Closes https://github.com/pytorch/pytorch/issues/28881

Differential Revision: [D20020183](https://our.internmc.facebook.com/intern/diff/D20020183/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20020183/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -439,7 +439,28 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
             --serverActiveCalls_;
             if (!futureResponse->hasError()) {
-              send(work.from_, std::move(*futureResponse).moveValue());
+              // Hack: Check for device type here since throwing an exception in
```
Would I be correct if I assume we wouldn't reach here at all in normal cases as we have already errored out in `wireSerialize`?

pritamdamania87 replied:
No we would reach here in normal cases if the response has a GPU tensor.

mrshenli replied:
Oh I see, it is for response. In that case shall we use the same check in `wireSerialize` and convert the error into an error message in `handleSend` for the response? So that we don't need to have the same logic duplicated in two places. 

Just to check IIUC, we do it here because this is on a separate thread so that we don't have the opportunity to directly throw the error, and hence explicitly check it and turn it into an error message?

pritamdamania87 replied:
Good point, let me try to merge it.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -439,7 +439,28 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+                  errorMsg = c10::str(
+                      "ProcessGroup RPC backend only supports",
+                      " CPU tensors, please move your tensors to CPU before sending ",
+                      "them over RPC. Found tensor on device: ",
```
Shall we make this error message different from the one in `wireSerialize`? If we got here, it seems we hit a bigger problem.

pritamdamania87 replied:
Its a similar issue as wireSerialize, the only difference here is that the Tensor is being sent as a response to an RPC.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1652,3 +1653,78 @@ def test_function_not_on_callee(self):
+    sys.version_info < (3, 0),
+    "Pytorch distributed rpc package " "does not support python2",
+)
+class RpcJitTest(RpcAgentTestFixture):
```
JIT tests are in `rpc/jit/*`?

pritamdamania87 replied:
Must be a bad rebase, will remove this.

mrshenli reviewed (commented):
Shall we add a warning to our API doc highlighting this requirement?

Besides, we also need to modify our tutorials/examples accordingly I think.

pritamdamania committed:
Update on "Disallow sending CUDA tensors over RPC for current RPC agents."


For our current RPC agents, this PR disallows sending CUDA tensors
over RPC and asks users to copy them explicitly to CPU. Currently, this seems
to be the easiest contract to guarantee for our current RPC agents, otherwise
if we do support this transparently it gets a little tricky in terms of whether
a CUDA tensor on the client should be sent to CPU/GPU of the remote end and
also which GPU device on the remote end.

In the future, the TensorPipe RPC agent can have its own specific handling of
CUDA tensors.

Closes https://github.com/pytorch/pytorch/issues/28881

Differential Revision: [D20020183](https://our.internmc.facebook.com/intern/diff/D20020183/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20020183/)!

[ghstack-poisoned]

pritamdamania87 commented:
@mrshenli I added a warning to our API documents. Regarding tutorials/examples, I'd prefer updating them in a separate PR. Do you know which tutorials/examples would need changes? I didn't see any GPU tensors being used in this tutorial: https://pytorch.org/tutorials/intermediate/rpc_tutorial.html

mrshenli reviewed (commented):
Just check the example looks fine. We already explicitly move the input to cuda and output back to cpu in the forward function:

https://github.com/pytorch/examples/blob/f692b1f16bf03129338442c3e66d342e28c34f31/distributed/rpc/rnn/rnn.py#L37-L48

pritamdamania committed:
Update on "Disallow sending CUDA tensors over RPC for current RPC agents."


For our current RPC agents, this PR disallows sending CUDA tensors
over RPC and asks users to copy them explicitly to CPU. Currently, this seems
to be the easiest contract to guarantee for our current RPC agents, otherwise
if we do support this transparently it gets a little tricky in terms of whether
a CUDA tensor on the client should be sent to CPU/GPU of the remote end and
also which GPU device on the remote end.

In the future, the TensorPipe RPC agent can have its own specific handling of
CUDA tensors.

Closes https://github.com/pytorch/pytorch/issues/28881

Differential Revision: [D20020183](https://our.internmc.facebook.com/intern/diff/D20020183/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20020183/)!

[ghstack-poisoned]

jjlilley commented:
btw, LGTM, though still seems to need to rebase - the Phabricator-side part seems to show a merge conflict. I'm fine pending that/tests working.

pritamdamania committed:
Update on "Disallow sending CUDA tensors over RPC for current RPC agents."


For our current RPC agents, this PR disallows sending CUDA tensors
over RPC and asks users to copy them explicitly to CPU. Currently, this seems
to be the easiest contract to guarantee for our current RPC agents, otherwise
if we do support this transparently it gets a little tricky in terms of whether
a CUDA tensor on the client should be sent to CPU/GPU of the remote end and
also which GPU device on the remote end.

In the future, the TensorPipe RPC agent can have its own specific handling of
CUDA tensors.

Closes https://github.com/pytorch/pytorch/issues/28881

Differential Revision: [D20020183](https://our.internmc.facebook.com/intern/diff/D20020183/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20020183/)!

[ghstack-poisoned]
